naming-conventions.md

```diff
@@ -0,0 +1,106 @@
+
+### Polymorphic Types
+
+Prefer short, single-letter names for polymorphic type variables and arguments. You'll notice a number of conventions with type variables.
+
+Unconstrained type variables often start with `a` and continue alphabetically
+```haskell
+id :: a -> a
+const :: a -> b -> a
+```
+
+Polymorphic lists are often "pluralized" by appending `s`
+```haskell
+filter :: (a -> Bool) -> [a] -> [a]
+filter _ [] = []
+filter predicate (a:as)
+  | predicate a = a : bs
+  | otherwise = bs
+ where
+  bs = filter predicate as
+```
+
+`f` is for `Functor` and sometimes `Applicative` and `Alternative`
+```haskell
+fmap :: Functor f => (a -> b) -> f a -> f b
+liftA2 :: Applicative f => (a -> b -> c) -> f a -> f b -> f c
+guard :: Alternative f => Bool -> f ()
+```
+
+`t` might be `Traversable` or `Foldable`
+```haskell
+foldr :: Foldable t => (a -> b -> b) -> b -> t a -> b
+traverse :: (Applicative f, Traversable t) => (a -> f b) -> t a -> f (t b)
+```
+
+`m` is always always `Monad`
+```haskell
+when :: Monad m => Bool -> m () -> m ()
+```
+
+`s` is for state and `e` is for environment
+```haskell
+put :: MonadState s m => s -> m ()
+ask :: MonadReader e m => m e
+```
+
+These are conventions only, but picking something unusual for a common typeclass might surprise other readers.
+
+### Concrete Types
+
+Avoid using single-letter names or abbreviations for values with a concrete type, especially if that type is domain-specific.
+
+For example, don't do this:
+```haskell 
+fetchCourseMemberships :: MonadIO m => Entity Teacher -> SqlReadT m [Entity CourseMembership]
+fetchCourseMemberships (Entity tId teacher) =
+  select $ from $ \(ts `InnerJoin` cs `InnerJoin` cms) -> do
+    on $ cs ^. CourseId ==. cms  ^. CourseMembershipCourseId
+    on $ cs ^. CourseTeacherId ==. ts ^. TeacherId
+    where_ $
+      case teacherSchoolId teacher of
+        Just sId ->
+          ts ^. TeacherSchoolId ==. val (Just sId)
+        Nothing ->
+          ts ^. TeacherId ==. val tId
+    pure cms
+```
+
+Do this instead:
+```haskell
+fetchCourseMemberships :: MonadIO m => Entity Teacher -> SqlReadT m [Entity CourseMembership]
+fetchCourseMemberships (Entity teacherId teacher) =
+  select $ from $ \(teachers `InnerJoin` courses `InnerJoin` memberships) -> do
+    on $ courses ^. CourseId ==. memberships  ^. CourseMembershipCourseId
+    on $ courses ^. CourseTeacherId ==. teachers ^. TeacherId
+    where_ $
+      case teacherSchoolId teacher of
+        Just schoolId ->
+          teachers ^. TeacherSchoolId ==. val (Just schoolId)
+        Nothing ->
+          teachers ^. TeacherId ==. val teacherId
+    pure memberships
+```
+
+One abbreviation we seem to tolerate is `num` as in `numQuestionsAnswered`. There may be others.
```

Comment:
Eh, I don't know if we should explicitly say this because naming it `numQuestionsAnswered` is better imo

Reply:
"tolerate" might be the issue here. Instead maybe we should turn this into a list of accepted abbreviations.

Reply:
Sorry, this was unclear. I wasn't saying `num` by itself is tolerated (I don't think I've ever seen that), just that it's tolerated _in identifiers_

Reply:
Oooooh, I see. That was a little unclear haha